metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala

```diff
@@ -0,0 +1,390 @@
+          .withTransformFetch(Some { fetch =>
+            fetch.addRepositories(
+              coursierapi.MavenRepository
+                .of("https://dl.bintray.com/alexarchambault/metals-amm")
+            )
+          })
+          .command() match {
+          case Left(e) =>
+            scribe.error(
+              s"Error getting Ammonite ${versions.ammoniteVersion} (scala ${versions.scalaVersion})",
+              e
+            )
+            Left(e)
+          case Right(command) =>
+            lastImportVersions = VersionsOption(
+              Some(versions.ammoniteVersion),
+              Some(versions.scalaVersion)
+            )
+            Right((command, path))
+        }
+      case Some(path) =>
+        val msg =
+          if (path.toNIO.getFileName.toString.endsWith(".worksheet.sc"))
+            "Current document is a worksheet, not an Ammonite script"
+          else
+            "Current document is not an Ammonite script"
+        scribe.error(msg)
+        Left(new Exception(msg))
+    }
+
+    commandScriptOpt match {
+      case Left(e) =>
+        val f = new CompletableFuture[Object]
+        f.completeExceptionally(e)
+        f
+      case Right((command, script)) =>
+        val extraScripts = buffers.open.toVector
+          .filter(_.isAmmoniteScript)
+          .filter(_ != script)
+        val futureConn = BuildServerConnection.fromSockets(
+          workspace(),
+          buildClient(),
+          languageClient(),
+          () =>
+            Ammonite
+              .socketConn(command, script +: extraScripts, workspace()),
+          tables().dismissedNotifications.ReconnectAmmonite
+        )
+        val f = for {
+          conn <- futureConn
+          _ <- connectToNewBuildServer(conn)
+        } yield ()
+        f.asJavaObject
+    }
+  }
+
+  def stop(): CompletableFuture[Object] = {
+    lastImportVersions = VersionsOption(None, None)
+    disconnectOldBuildServer().asJavaObject
+  }
+
+  def generatedScalaPath(
+      targetId: BuildTargetIdentifier,
+      source: AbsolutePath
+  ): Option[AbsolutePath] =
+    if (Ammonite.isAmmBuildTarget(targetId) && source.isAmmoniteScript)
+      buildTargets.scalacOptions(targetId).map { target =>
+        val rel = source.toRelative(workspace())
+        val path = Paths
+          .get(new URI(target.getClassDirectory))
+          .getParent
+          .resolve(
+            s"src/ammonite/$$file/${rel.toString.stripSuffix(".sc")}.scala"
+          )
+        AbsolutePath(path.toAbsolutePath.normalize)
+      }
+    else
+      None
+
+  def generatedScalaInputForPc(
+      targetId: BuildTargetIdentifier,
+      source: AbsolutePath,
+      position: Position
+  ): Option[(Input.VirtualFile, Position)] =
+    generatedScalaPath(targetId, source)
+      .map { scalaPath =>
+        val scInput = source.toInputFromBuffers(buffers)
+        val input = scalaPath.toInput
+
+        val updatedContent = input.value
```

Comment:
Could we document how do we generate the source files for PC?

Reply:
Done.